### DISCUSSION TITLE
Confused on how to solve Error: The edge runtime does not support Node.js 'crypto' module after adding middleware.ts

### DISCUSSION THREAD

**Discussion by haocloo** [vercel/next.js#62985] [Help] [answered]
### Summary

May I know how to solve this issue of Error: The edge runtime does not support Node.js 'crypto' module? This error happened  after I added the middleware.ts file. 

this my .env.local
```bash
POSTGRES_URL=
APP_ENV=local

# Must follow naming convention for NextAuth v5 to work
AUTH_SECRET=

AUTH_GITHUB_ID=
AUTH_GITHUB_SECRET=

AUTH_GOOGLE_ID=
AUTH_GOOGLE_SECRET=
```

middleware.ts
```ts
import { auth } from "@/lib/auth/auth"

export default auth((req) => {
  // req.auth
  console.log("Route Middleware", req.nextUrl.pathname)
})

// Optionally, don't invoke Middleware on some paths
// Read more: https://nextjs.org/docs/app/building-your-application/routing/middleware#matcher
export const config = {
// if you dont match anything, the website will work.
    matcher: ["/((?!api|_next/static|_next/image|favicon.ico).*)"],
    // matcher: ['/((?!.+\\.[\\w]+$|_next).*)', '/', '/(api|trpc)(.*)'],
}
```

the file at @/lib/auth/auth
```ts
// auth.ts
import NextAuth from "next-auth"
import GitHub from "next-auth/providers/github"
import Google from "next-auth/providers/google"
import CredentialsProvider from "next-auth/providers/credentials";
import { db } from "@/lib/db/index"
import { eq } from "drizzle-orm";
import { getUserByEmail } from "@/data/user";
import { users } from "../db/schema";

let userInfo = {
  id: "",
  email: "",
  name: "",
}

// sign in with email and password
async function authorize(credentials: any) {
  if (!credentials?.email || !credentials.password) {
    return null;
  }
  
  let user = await getUserByEmail(credentials.email as string);
  
  if (!user?.length) {
    return null;
  }
  
  userInfo = {...user as any};
  
  return user;
}

// sign in with google or github
async function signIn({ user, account, credentials }: { user: any; account: any; credentials: any }) {
  // prevent signInUser from running if authorizeUser has already run
  if(credentials?.email){
    return true;
  }
  if (account?.provider === "google" || account?.provider === "github") {
    try {
      if(user?.email){
        let result = await getUserByEmail(user.email);
        if (result?.length === 0) {
          await db.insert(users).values({ name: user.name as string ,email: user.email as string, password: account.provider as string });
          result = await getUserByEmail(user.email);
        }

        userInfo = {...result as any};

        return true;
      }
    } catch (err) {
      console.log(err);
      return false;
    }
  }
  return false;
}

async function jwt({ token, user, account }: { token: any; user: any; account: any }) {
  // cant use the user object to get the id cus the data is not updated when using oauth but only in credentials login
  // user object contains info from credentials or from oauth default user object, it will then disappaer aftre this function runs
  if (user || account?.provider === "google" || account?.provider === "github") {
    return userInfo;
  }  
  return token;
}

async function session({ session, token }: { session: any; token: any }) {
  if (token) {
    session.user = token;
  }
  return session.user;
}

export const {
  handlers: { GET, POST },
  auth,
  // signIn,
  signOut,
} = NextAuth({
  secret: process.env.AUTH_SECRET,
  providers: [GitHub,Google,
    CredentialsProvider({
      name: "Credentials",
      credentials: {
        email: {
          label: "Email",
          type: "email",
          placeholder: "example@example.com",
        },
        password: { label: "Password", type: "password" },
      },
      authorize
    }),
  ],
  pages : {
    signIn: '/login',
  },
  callbacks: {
    signIn,
    jwt,
    session
  }
})
```

this is the file at app/api/auth/[...nextauth]/route.ts
```ts
export { GET, POST } from "@/lib/auth/auth";
```

This is the error logged in the terminal.
```bash
⚠ ./node_modules/.pnpm/pg@8.11.3/node_modules/pg/lib/native/client.js
Module not found: Can't resolve 'pg-native' in 'C:\Users\User\Desktop\webwebweb\food-order\node_modules\.pnpm\pg@8.11.3\node_modules\pg\lib\native'

Import trace for requested module:
./node_modules/.pnpm/pg@8.11.3/node_modules/pg/lib/native/client.js
./node_modules/.pnpm/pg@8.11.3/node_modules/pg/lib/native/index.js 
./node_modules/.pnpm/pg@8.11.3/node_modules/pg/lib/index.js        
./lib/db/index.ts
./lib/auth/auth.ts
 ⨯ Error: The edge runtime does not support Node.js 'crypto' module.
Learn More: https://nextjs.org/docs/messages/node-module-in-edge-runtime
    at <unknown> (webpack-internal:///(middleware)/./node_modules/.pnpm/next@14.1.2_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/esm/server/web/globals.js:33)
    at Object.get (webpack-internal:///(middleware)/./node_modules/.pnpm/next@14.1.2_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/esm/server/web/globals.js:33:19)
    at eval (webpack-internal:///(middleware)/./node_modules/.pnpm/pg@8.11.3/node_modules/pg/lib/crypto/utils-webcrypto.js:22:32)     
    at (middleware)/./node_modules/.pnpm/pg@8.11.3/node_modules/pg/lib/crypto/utils-webcrypto.js (file://C:\Users\User\Desktop\webwebweb\food-order\.next\server\middleware.js:1251:1)
    at __webpack_require__ (file://C:\Users\User\Desktop\webwebweb\food-order\.next\server\edge-runtime-webpack.js:37:33)
    at fn (file://C:\Users\User\Desktop\webwebweb\food-order\.next\server\edge-runtime-webpack.js:285:21)
    at eval (webpack-internal:///(middleware)/./node_modules/.pnpm/pg@8.11.3/node_modules/pg/lib/crypto/utils.js:8:20)
    at (middleware)/./node_modules/.pnpm/pg@8.11.3/node_modules/pg/lib/crypto/utils.js (file://C:\Users\User\Desktop\webwebweb\food-order\.next\server\middleware.js:1262:1)
    at __webpack_require__ (file://C:\Users\User\Desktop\webwebweb\food-order\.next\server\edge-runtime-webpack.js:37:33)
    at fn (file://C:\Users\User\Desktop\webwebweb\food-order\.next\server\edge-runtime-webpack.js:285:21)
    at eval (webpack-internal:///(middleware)/./node_modules/.pnpm/pg@8.11.3/node_modules/pg/lib/crypto/sasl.js:3:16) {
  middleware: true
}
```

### Example

https://github.com/haocloo/issue

**Comment by andersonsoa** [comment]
same issue here =/

**Comment by icyJoseph** [comment]
You have to use the `middleware` version of next-auth: https://next-auth.js.org/configuration/nextjs#middleware

**Comment by Babailan** [comment]
because the runtime you are using is edge runtime which is not a node js server so crypto module is not available im also having this issue right now but i don't know to solve this

**Reply by icyJoseph** [reply]
There are libraries out there that use the Web Crypto, like `jose`, but it depends on your use case, what is it that you need?

**Reply by Babailan** [reply]
I'm using the Node.js MongoDB driver, which depends on the native crypto library, I'm pretty sure the errors are because of that. So, I'm thinking about switching to MongoDB Atlas's HTTP data API for better compatibility with edge runtimes.

**Reply by icyJoseph** [reply]
yeah, pretty much, when it is a dependency so nested that you can't do anything about it, it is better to find an alternative compatible with edge.

**Reply by Feeco-Li** [reply]
> I'm using the Node.js MongoDB driver, which depends on the native crypto library, I'm pretty sure the errors are because of that. So, I'm thinking about switching to MongoDB Atlas's HTTP data API for better compatibility with edge runtimes.

this actually explains where the problem is and how to solve this. :1st_place_medal:

**Comment by aliallam98** [comment]
i encountered this issue and the solution come from this file middleware
instead of  import { auth } from "@/auth"; like that try this way 
 

first thing import NextAuth from "next-auth";

import authConfig from "./auth.config"; this file in my case contains  the following : 
import GitHub from "next-auth/providers/github"

import type { NextAuthConfig } from "next-auth"

export default {
  providers: [GitHub],
} satisfies NextAuthConfig

then make that : const { auth } = NextAuth(authConfig);

it should works now  i hope you find this useful

**Reply by fristyr** [reply]
Doesn't really worked for me, not sure why

**Comment by lukasver** [comment]
I had the same issue when updating from v4 but actually while this error is showing up, the actual cause was related with something different:

My `/lib/auth` file was importing a logger from  `/lib/logger` which relied on `headers` function from `next/headers` and other server only functions from a logger library.  

I think that since the `auth` function now works universally through client/server components, apparently this triggered some (code splitting?) issue. 

Removing the import and refactoring a little did the trick.

```
Import trace for requested module:
./node_modules/@google-cloud/logging/build/src/index.js
./src/lib/loggers.ts. **<--- this import in the auth file was causing the problem** 
./src/lib/auth.ts
....
```

**Comment by gpoole** [comment]
If you are able to use the `jwt` session strategy instead of `database`, this guide shows you how to work around the error and get a working config even if you're using an adapter like MongoDB that relies on APIs that aren't edge-compatible:
https://authjs.dev/getting-started/migrating-to-v5#edge-compatibility

This should work fine for cases where you don't _need_ a `database` session strategy but [the config may have defaulted to one by adding a DB provider](https://authjs.dev/reference/core#strategy), like if you've followed the instructions to set up magic links.

**Comment by ChristianRich** [comment]
I solved this issue by not attempting to use MongoDB directly from the Next-Auth callback functions. It seems only Edge runtime is supported within this scope, meaning you can't use *many* internal node packages (including `crypto` which is used by MongoDB).

Instead I make an internal API route and invoke using `Fetch` from e.g the Next-Auth jwt callback when I want to either update an existing user's tokens or create a new user when they signup.

`src/auth.ts`

```ts
// Edge runtime. Limited support of built-in node module. E.g `crypto` is not supported and will fail at runtime.
import NextAuth from "next-auth";
import GoogleProvider from "next-auth/providers/google";

// Docs:
// https://next-auth.js.org/configuration/providers/oauth
export const {
    handlers: { GET, POST },
    auth,
    signIn,
    signOut,
} = NextAuth({
    providers: [
        GoogleProvider({
            clientId: process.env.GOOGLE_CLIENT_ID,
            clientSecret: process.env.GOOGLE_CLIENT_SECRET,
            allowDangerousEmailAccountLinking: true,
            authorization: {
                params: {
                    prompt: "consent",
                    access_type: "offline",
                    response_type: "code",
                },
            },
        }),
    ],
    // Docs
    // https://next-auth.js.org/configuration/callbacks
    callbacks: {
        async jwt({ trigger, token, account, profile, user }) {
            if (trigger === 'signUp') {
                try {
                    const response = await fetch(`${process.env.NEXTAUTH_URL}/api/user`, {
                        method: "POST",
                        headers: {
                            "Content-Type": "application/json",
                        },
                        body: JSON.stringify({token, account, profile, user}),
                    });
    
                    if (!response.ok) {
                        console.error("Failed to create user:", await response.json());
                    }
                } catch (error) {
                    console.error("Error calling API:", error);
                }
            }

            return token;
        },
    },
});
```

Then in the route handler, I can safely do my DB instantiation and a variety of operations using the node.js runtime rather than Edge.

`src/app/api/user/route.ts`

```ts
// Nodejs runtime. Full support of node.js runtime. Do what you want.
import { connectMongoDB } from '@/lib/common/mongodb/connect';
import { UserQuery } from '@/lib/services/mongodb/user-query';
import { NextResponse } from 'next/server';

export async function POST(req: Request) {
    try {
        await connectMongoDB(); // Singleton style connector
        const body = await req.json();
        const user: any = await UserQuery.getInstance().create(body);
        return NextResponse.json({ status: 'success', data: { id: user.id} });
    } catch (error) {
        console.error(error);
        return NextResponse.json(
            { status: 'error', message: 'Failed to create user' },
            { status: 500 }
        );
    }
}
```

**Reply by vpaflah** [reply]
But, the endpoint will be public and everyone can access it.

**Reply by monteiz** [reply]
@vpaflah you can pass the `AUTH_SECRET` and return `404` if it is not correct.

``` js
// src/app/api/user/route.ts
export async function POST(req) {
	const { email, password, secret } = await req.json();

	if (secret !== process.env.AUTH_SECRET) {
		return NextResponse.json({}, { status: 404 }); // going stealth mode...
	}
	
	// verify the user and return the result
}
```

**Comment by wilsonllucena** [comment]
I solved it this way

```javascript
// middleware.ts

import { cookies } from "next/headers";
import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";

const protectedPaths = [
  "/dashboard",
];

export async function middleware(request: NextRequest) {
  const path = request.nextUrl.pathname;

  const isProtectedPath = protectedPaths.some((prefix) => 
    path.startsWith(prefix)
  );

  if (isProtectedPath) {
    const token = (await cookies()).get('authjs.session-token');
    if (!token) {
      const url = new URL("/", request.url);
      return NextResponse.redirect(url);
    }
  }

  return NextResponse.next();
}

// Configure matcher to specify which paths middleware will run on
export const config = {
  matcher: [
    "/((?!api|_next|static|favicon.ico|sitemap.xml).*)",
  ],
};
```

**Reply by fulcanelly** [reply]
it's super shitty solution 

this way you have zero security 

you dont validate token details you just check presence of token

**Reply by suchi-mayvenr** [reply]
totally agreed, its only enough to check whether the user is signed in or not. trying to implement any form of RBAC would be impossible.

**Comment by littlezabi** [comment]
The same error occurs when I'm trying to validate my JSON Web Token. The alternative is the jose library, but I don't want to use it because I prefer not to install too many libraries in my project. I'm trying hard to fix the issue, so if anyone knows a solution, please help.

**Comment by leerob** [comment, marked as answer]
Good news everyone, **Node.js Middleware support landed in Next.js 15.2** (experimental). This is now ready for testing. For any issues you run into, please open new issues so we can track and address before marking it as stable. Thank you!

https://nextjs.org/blog/next-15-2#nodejs-middleware-experimental

**Reply by D50000** [reply]
I will try it later, TY.

**Reply by Cranbaerry** [reply]
Hey, I'm still having this issue despite having the experimental option toggled on. What gives?
![image](https://github.com/user-attachments/assets/cf49821b-dc90-4efb-9dc2-56f38d647cc2)

**Reply by TechQuery** [reply]
@leerob When I try to use `jsonwebtoken` package in App router middlewares, I failed with this error:

> [CanaryOnlyError: The experimental feature "experimental.nodeMiddleware" can only be enabled when using the latest canary version of Next.js.]

**Reply by vpaflah** [reply]
> @leerob When I try to use `jsonwebtoken` package in App router middlewares, I failed with this error:
> 
> > [CanaryOnlyError: The experimental feature "experimental.nodeMiddleware" can only be enabled when using the latest canary version of Next.js.]

You must install the canary version of nextjs to be able to use the experimental feature.

**Reply by suryavirkapur** [reply]
You have to put the 

```ts
export const config = {
  runtime: 'nodejs',
};
```


part.

**Reply by iamCoderMohit** [reply]
this worked for me

export const config = {
  runtime: 'nodejs',
};

**Comment by ShayokhShorfuddin** [comment]
Hello. I am using Authjs, and my Next.js version is `^15.5.2`. Even though Node.js Middleware support was added in Next.js 15.2, I still can't seem to use it.

<img width="834" height="538" alt="image" src="https://github.com/user-attachments/assets/6206dc39-2c3f-49fa-8211-cfc9381af9fd" />
<img width="992" height="453" alt="image" src="https://github.com/user-attachments/assets/ecc1c2e7-7225-4847-932c-29c5371dfd39" />
<img width="423" height="156" alt="image" src="https://github.com/user-attachments/assets/d94d4b6d-f096-4134-afff-6693cd398976" />
<img width="769" height="662" alt="image" src="https://github.com/user-attachments/assets/45d384a9-71eb-494f-9bc2-54d85841fb33" />
<img width="1073" height="821" alt="image" src="https://github.com/user-attachments/assets/a4f8e555-5e6d-4bc0-9a1c-284093ddbea9" />

**Reply by icyJoseph** [reply]
15.5 stabilized nodeMiddleware https://nextjs.org/docs/app/api-reference/file-conventions/middleware#runtime

You don't need a `next.config.js` flag anymore, all you gotta do is add the runtime config to `middleware`:

```ts
export const config = {
  runtime: 'nodejs',
}
```

**Reply by ShayokhShorfuddin** [reply]
> 15.5 stabilized nodeMiddleware https://nextjs.org/docs/app/api-reference/file-conventions/middleware#runtime
> 
> You don't need a `next.config.js` flag anymore, all you gotta do is add the runtime config to `middleware`:
> 
> ```ts
> export const config = {
>   runtime: 'nodejs',
> }
> ```

I just added a `middleware.ts` and exported the config. Also removed the flag from the `next.config.ts`. However, the issue persists when I try to visit `/home` route which is added in the `matcher`.

<img width="898" height="939" alt="image" src="https://github.com/user-attachments/assets/16d86d14-5e31-47c1-9ae2-fee4128770c8" />
<img width="790" height="933" alt="image" src="https://github.com/user-attachments/assets/0f53daad-c173-4518-9d7e-b3e4a22f6889" />

**Reply by ddaoxuan** [reply]
> Hello. I am using Authjs, and my Next.js version is `^15.5.2`. Even though Node.js Middleware support was added in Next.js 15.2, I still can't seem to use it.
> 
> <img alt="image" width="834" height="538" src="https://private-user-images.githubusercontent.com/56217851/483925962-6206dc39-2c3f-49fa-8211-cfc9381af9fd.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcwODI2MzQsIm5iZiI6MTc1NzA4MjMzNCwicGF0aCI6Ii81NjIxNzg1MS80ODM5MjU5NjItNjIwNmRjMzktMmMzZi00OWZhLTgyMTEtY2ZjOTM4MWFmOWZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTA1VDE0MjUzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ3YmE4NDM4ZmEyMzgyN2VhMGJiYWFjODkzYzJmNzQ3ODlkNTFjYjlmZjJjZGI2N2YxNWU5MDM4OTAzZGNjNTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lWA5ugCvFeJp-WPoudEppMVayqGQnFW9yFShUpH8DF4"> <img alt="image" width="992" height="453" src="https://private-user-images.githubusercontent.com/56217851/483925966-ecc1c2e7-7225-4847-932c-29c5371dfd39.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcwODI2MzQsIm5iZiI6MTc1NzA4MjMzNCwicGF0aCI6Ii81NjIxNzg1MS80ODM5MjU5NjYtZWNjMWMyZTctNzIyNS00ODQ3LTkzMmMtMjljNTM3MWRmZDM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTA1VDE0MjUzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc2MmU4MTc4MjM0ZDIwMzQ4MjBlYmJmMzY0Nzc4OGNlYWQ0Y2Y2MjU4YzEzNDgxYWI1MGVkMzA2N2Q2Njc4M2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.zlNYd_i7Rw2qck2Yfyolj6bbVbTmPy8uR0MRsfLRbzs"> <img alt="image" width="423" height="156" src="https://private-user-images.githubusercontent.com/56217851/483925968-d94d4b6d-f096-4134-afff-6693cd398976.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcwODI2MzQsIm5iZiI6MTc1NzA4MjMzNCwicGF0aCI6Ii81NjIxNzg1MS80ODM5MjU5NjgtZDk0ZDRiNmQtZjA5Ni00MTM0LWFmZmYtNjY5M2NkMzk4OTc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTA1VDE0MjUzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwM2M0MGU5M2VlYTk5ZmRiMjNiY2E4N2QzMmIyNjUyZjgxMWIwZTdiNDg3ZGNkMDk2ZGFmNjJlOWY4YjJkYzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MQQVkvcmzIaHKwiwyD7j3yTzKaxUda8_zWkhRP3JgZU"> <img alt="image" width="769" height="662" src="https://private-user-images.githubusercontent.com/56217851/483925975-45d384a9-71eb-494f-9bc2-54d85841fb33.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcwODI2MzQsIm5iZiI6MTc1NzA4MjMzNCwicGF0aCI6Ii81NjIxNzg1MS80ODM5MjU5NzUtNDVkMzg0YTktNzFlYi00OTRmLTliYzItNTRkODU4NDFmYjMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTA1VDE0MjUzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjYzIxZjdmYTVhYzMzZTNmMjczOGU4YTVhNmRlZjllZGZmNjJhMzI4Njc1YWIyMGQxNDk2NTA1MGUzYmQyNjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.50JDNk94MHP6JyQhfAAU0L3YSvRR237Iqu8dA7W8JKk"> <img alt="image" width="1073" height="821" src="https://private-user-images.githubusercontent.com/56217851/483925978-a4f8e555-5e6d-4bc0-9a1c-284093ddbea9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcwODI2MzQsIm5iZiI6MTc1NzA4MjMzNCwicGF0aCI6Ii81NjIxNzg1MS80ODM5MjU5NzgtYTRmOGU1NTUtNWU2ZC00YmMwLTlhMWMtMjg0MDkzZGRiZWE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTA1VDE0MjUzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkyMTJlYWNkMzI1ODhmZmExNzRhYTkzYjNjOGQ2YWI2MjU5Mzk2MTM0MWY4MjRkNTBmNjliNzQ0NzgzMjBmYWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ilONPgJaznzOdFWG4QAEyhSVPoc8_MjGpnS5ebaQnMk">

Hey. I  think you are very likely affected by this bug too. I just faced the same thought might be worth sharing:

https://github.com/vercel/next.js/issues/83474

when removed turbopack now my config is respected and it seems to be using node runtime instead of edge.

So to fix you'd have to remove turbopack for now from your scripts and stop using it unless fixed

**Reply by icyJoseph** [reply]
This has been fixed in canary, [v15.5.1-canary.27](https://github.com/vercel/next.js/releases/tag/v15.5.1-canary.27) - next stable version should have it

**Reply by ShayokhShorfuddin** [reply]
I upgraded to next 15.5.3 but still getting the same issue when running on npm `dev` and `build`.

https://github.com/user-attachments/assets/92c7e8c8-fd25-4f28-9d08-122c1d564a2c

**Reply by icyJoseph** [reply]
And the runtime config? 

- https://nextjs.org/docs/app/api-reference/file-conventions/middleware#runtime

```ts
export const config = {
  matcher: ["/nome/:path"],
  runtime: 'nodejs' // <--- you are missing this!
}
```

You only had a `matcher`. 

And since you are using Turbopack, in 15.5.3 the config is not being picked up correctly. Canary v15.5.1-canary.x is still ahead of 15.5.3 (I know it is confusing)

**Reply by ShayokhShorfuddin** [reply]
Ohh yeah, silly me. 😖 Thanks a lot

**Comment by gpiechnik2** [comment]
Add:

```
export const runtime = "nodejs";
```

on the top of your `middleware.ts` file. It works for me
